fdbserver/ptxn/TLogServer.actor.cpp

```diff
@@ -1115,76 +1124,71 @@ ACTOR Future<Void> tLogCommit(Reference<TLogGroupData> self,
 	return Void();
 }
 
-Optional<std::pair<Version, StringRef>> LogGenerationData::getSerializedTLogData(const Version& version,
-                                                                                 const StorageTeamID& storageTeamID) {
+Optional<std::pair<Version, std::pair<StringRef, Arena>>> LogGenerationData::getSerializedTLogData(
+    const Version& version,
+    const StorageTeamID& storageTeamID) {
 
 	auto pStorageTeamData = getStorageTeamData(storageTeamID);
 	// by lower_bound, if we pass in 10, we might get 12, and return 12
 	auto iter = pStorageTeamData->versionMessages.lower_bound(version);
 	if (iter == pStorageTeamData->versionMessages.end()) {
-		return Optional<std::pair<Version, StringRef>>();
+		return Optional<std::pair<Version, std::pair<StringRef, Arena>>>();
 	}
 
-	return std::make_pair(iter->first, iter->second.first);
+	return std::make_pair(iter->first, iter->second);
 }
 
 static const size_t TLOG_PEEK_REQUEST_REPLY_SIZE_CRITERIA = 1024 * 1024;
 
-ACTOR Future<Void> tLogPeekMessages(TLogPeekRequest req, Reference<LogGenerationData> logData) {
-	ASSERT(logData.isValid());
-
-	wait(logData->version.whenAtLeast(req.beginVersion));
-
-	if (!logData->getStorageTeamData(req.storageTeamID).isValid()) {
-		req.reply.sendError(storage_team_id_not_found());
-		return Void();
+Version poppedVersion(Reference<LogGenerationData> logData, StorageTeamID teamID) {
+	auto teamData = logData->getStorageTeamData(teamID);
+	if (!teamData) {
+		if (teamID == txsTeam) {
+			return 0;
+		}
+		return logData->recoveredAt + 1;
 	}
+	return teamData->popped;
+}
 
-	TLogPeekReply reply;
-	TLogSubsequencedMessageSerializer serializer(req.storageTeamID);
+void peekMessagesFromMemory(const TLogPeekRequest& req,
+                            std::vector<std::pair<Version, std::pair<StringRef, Arena>>>& values,
```

Comment:
Nit: change to pointer type to signify it's mutable, since this holds the return value.